2.0.0-Data-Model.md

```diff
@@ -12,22 +12,84 @@ We are starting with a read-only API due to potentially different views and appr
-
-For example, if a WES server was asked to process `trs://trs.example.org/123456323`, it would know that it could issue a GET request to `https://trs.example.org/api/ga4gh/trs/v2/tools/123456323` to learn how to fetch that object.
+#### TRS Tool and TRS Tool Version IDs
+
+Each implementation of TRS can choose its own identifier scheme, as long as it
+follows these guidelines:
+
+* TRS Tool and TRS Tool Version IDs are strings made up of uppercase and
+  lowercase letters, decimal digits, hyphen, period, underscore and tilde
+  (`[A-Za-z0-9.-_~]`). See [RFC 3986 §
+  2.3](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3). TRS Tool and
+  TRS Tool Version IDs MAY further contain percent characters (`%`) whenever
+  they are used in API calls, but only if they were introduced through
+  percent-encoding of any non-valid characters (see next bullet point).
+* TRS Tool and TRS Tool Version IDs can contain other characters, but they MUST
+  be percent-encoded (see [RFC 3986 §
+  2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4)) into valid
+  TRS Tool and TRS Tool Version IDs as per the previous rule whenever they are
+  used in API calls. This is because non-encoded IDs may interfere with the
+  interpretation of routes, e.g., for the `/tools/{id}/versions/{version_id}`
+  endpoint.
+* Any given TRS Tool or TRS Tool Version ID MUST always identify the same
+  resource (tool or tool version, respectively) on a given TRS implementation.
+  This constraint aids with reproducibility.
+* TRS implementations MAY have more than one TRS Tool or TRS Tool Version ID
+  mapping to the same resource (tool or tool version, respectively).
 
-This recommendation is intended to mirror the discussion that went into the [DRS URI scheme](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_uris).
+#### TRS URIs
 
-For informational purposes, we recommend that TRS implementations add themselves to https://github.com/ga4gh/tool-registry-service-schemas/blob/develop/registry.json to provide for the possibility of creating a global indexing service and to allow others to more easily discover a TRS implementation. 
+To conveniently pass content references to TRS resources, e.g., to advise a
+[WES](https://github.com/ga4gh/workflow-execution-service-schemas) or
+[TES](https://github.com/ga4gh/task-execution-schemas) service which tool
+(version) to use, we define a URI syntax for TRS-accessible content. Strings of
+the form `trs://<server>/<id>` (unversioned) and
+`trs://<server>/<id>/<version_id>` (versioned) mean _“you can fetch the content
+with TRS Tool ID `<id>` and, if provided, TRS Tool Version ID `<version_id>`
+from the TRS server at `<server>` "_.
+
+For example, if a TRS client was asked to process
+`trs://trs.example.org/tool_ABC/v1.2.3`, it would know that it could, e.g.,
+issue `GET` requests to
+`https://trs.example.org/api/ga4gh/trs/v2/tools/tool_ABC` and
```

Comment:
@uniqueg This example is a bit awkward and likely points to a deficiency in the current approach for defining trs identifiers (and potentially drs identifiers as well). The TES specification defines the root path of the API as `/ga4gh/trs/<version>`. 

Putting aside specific trs versioning issues (from the conversation above), I am now left with another dilema: TRS identifiers as defined only provide the `hostname` and a `toolId` and possibly a version `id`. This approach does not leave ANY room for path values that may come before the `/ga4gh/...` prefix. Given this, how am I supposed to resolve the actual trs URL for the example?

By convention, any tooling I create would resolve `trs://trs.example.org/tool_ABC/v1.2.3`  to `https://trs.example.org/ga4gh/trs/v2/tools/tool_ABC/versions/v1.2.3`, which is not what is presented in this example (note the missing API). I would not be so concerned over this if it was not already present in the wild in the Dockstore api (@denis-yuen). So my question then is how can we possibly resolve the path components without introducing ambiguity into the trs identifier.

One approach could be just adding the prefix paths prior to the `toolId trs://trs.example.org/api/tool_ABC`. Unfortuantely this is ambiguous since this could be referring to either a tool with the path prefix `api` and version `tool_ABC` or a tool `tool_ABC` with a path `api`. If we do not allow the path prefix to be in the identifier, any tooling is limited to the trs implementations that It already knows how to resolve, effectively limiting the usefulness of the trs id in general. 

We essentially need a reliable way to extract the `toolId` and `versionId` without any ambiguity, while still allowing preceding fragments in the path. Some examples are below:

```
trs://trs.example.org/<prefix-before-ga4gh-api>/<toolId>:<versionId>
trs://trs.example.org/<prefix-before-ga4gh-api>/<toolId>/v/<versionId>
trs://trs.example.org/<prefix-before-ga4gh-api>/<toolId>/version/<versionId>
```

Reply:
One could probably calculate the prefixes on the fly from https://github.com/ga4gh/tool-registry-service-schemas/blob/develop/registry.json but it probably does make more sense to make the URI format since we're not using it yet(?)

Reply:
@denis-yuen I would advise against relying on prefixes in the `registry.json` file, which automatically enforces that file is the authoratative source of all tool registries. It does not really allow for new registries to dynamically be created (ie in a multitenant system)

Reply:
I think my personal preference for approaches would likely be: `trs://trs.example.org/<prefix-before-ga4gh-api>/<toolId>:<versionId>`. The side effect of this is that `:` is now a restricted character that cannot be used in the identifier itself, but I think that is a reasonable restriction

Reply:
tl;dr

I agree with @patmagee's suggestion provided that DRS folks agree that it will work for them as well.

---

Good catch @patmagee - thanks a lot! I had actually realized this issue before when implementing a [TRS client that understands TRS URIs](https://github.com/elixir-cloud-aai/TRS-cli) and having to introduce an optional `base_path` parameter so that it would work with Dockstore. It seems I had forgotten about it though. And I agree, the TRS URI uses its utility if this has to be known beforehand.

From what I can see, @patmagee's suggestion should do the trick.

But thinking this through a little more, maybe there are a couple of things that we should discuss outside of this issue before merging:
- To make sure that whatever is decided here has the perspective of aligning with whatever is decided for DRS URIs in the future, I think this discussion should include the DRS community. @patmagee's other worry of being able to include basically any character in TRS Tool/Version IDs could be brought up there as well.
- I am not quite sure whether API versioning could have any effect on a client's ability to properly resolve a TRS URI. I feel that it should not. But without explicit guidelines on API versioning, outlining expected behaviors and discussing their impact on reproducibility, I feel that it is hard to be sure. So perhaps it might be worth to discuss specifying such guidelines somewhere, possibly in a place that includes other work streams as well.

Besides that, it is possibly a good idea to aim for a consensus of how `servers` are defined in the various Cloud WS APIs in the mid to long term. And whatever is decided here, should not collide with our definitions of TRS and DRS URIs. Currently, the situation is like this:

**TRS**
```yaml
servers:
  - url: /ga4gh/trs/v2
```

**DRS**
```yaml
servers:
  - url: https://{serverURL}/ga4gh/drs/v1
    variables:
      serverURL:
        default: drs.example.org
        description: >
          DRS server endpoints MUST be prefixed by the '/ga4gh/drs/v1' endpoint
          path
```

**WES**
```yaml
servers:
- url: https://{defaultHost}/{basePath}/{apiVersion}
  variables:
    defaultHost:
      default: www.example.com
    basePath:
      default: ga4gh/wes
    apiVersion:
      default: v1
```

**TES**
```yaml
servers:
- url: /ga4gh/tes/v1
```

So, it looks like TRS and TES do not care where the API is deployed, as long as endpoint paths are immediately preceded by `/ga4gh/trs/v2` and `/ga4gh/tes/v1`, respectively. DRS is very similar but a little more prescriptive in that it requires the use of the `https` schema. WES, on the other hand, requires `https` as well, but otherwise only seems to require that there be at least two slashes included in the URL _before_ the endpoint path (weird?). In particular, it does not require a specific base path suffix like `/ga4gh/wes/v1` (though the default _does_ include it).

Now, there is currently no consensus across those definitions. But perhaps it seems to be going in the direction of the DRS definition, i.e., start with `https` and end with a specific base path suffix of the form `/ga4gh/{api_short_name}/{api_version}`. @patmagee's definition would account for that (having the client fill in the appropriate base path suffix). But even with a more relaxed definition that does not prescribed a specific base path for GA4GH Cloud APIs, the suggested solution should still work.

Reply:
Proposal at Cloud WS was to also break this into a separate ticket, but to prioritize it for discussion before 2.0.1 release and/or the GA4GH plenary 
https://github.com/ga4gh/tool-registry-service-schemas/issues/220